src/Model/User.php

```diff
@@ -515,7 +515,27 @@ public static function authenticate($user_info, $password, $third_party = false)
 	 */
 	public static function getIdFromPasswordAuthentication($user_info, $password, $third_party = false)
 	{
-		$user = self::getAuthenticationInfo($user_info);
+		// Addons registered with the "authenticate" hook may create the user on the
+		// fly. `getAuthenticationInfo` will fail if the user doesn't exist yet. If
+		// the user doesn't exist, we should give the addons a chance to create the
+		// user in our database, if applicable, before re-throwing the exception if
+		// they fail.
+		try {
+			$user = self::getAuthenticationInfo($user_info);
+		} catch (Exception $e) {
+			$username = (is_string($user_info) ? $user_info : $user_info['nickname'] ?? '');
+
+			// Addons can create users, and since this 'catch' branch should only
+			// execute if getAuthenticationInfo can't find an existing user, that's
+			// exactly what will happen here. Creating a numeric username would create
+			// abiguity with user IDs, possibly opening up an attack vector.
+			// So let's be very careful about that.
+			if (empty($username) || is_numeric($user_info) || is_numeric($user_info['nickname'] ?? '')) {
```

Comment:
`is_numeric($user_info)` already is checked on line 526, and `$user_info['nickname'] ?? ''` already is assigned to `$username`, so this condition can be simplified.
```suggestion
			if (empty($username) || is_numeric($username)) {
```

Reply:
Yup, good eye.